### PR TITLE
Loss

### PR DESCRIPTION
**Pull request by NRauschmayr**
I updated the loss tutorial with descriptions about currently supported loss functions.

### CODE REVIEW

File: python/guide/packages/gluon/loss.md [hunk 1]
```diff
@@ -79,38 +83,79 @@ def plot(x, y):
...
-    y = loss(x, nd.ones_like(x))
-    plot(x, y) 
-    
-show_classification_loss(gloss.LogisticLoss())
+show_regression_loss(gloss.HuberLoss(rho=1))
 ```
 
+#### [Cross Entropy Loss with Sigmoid](https://mxnet.incubator.apache.org/api/python/gluon/loss.html#mxnet.gluon.loss.SigmoidBinaryCrossEntropyLoss)
+
+Binary Cross Entropy is a loss function used for binary classification problems e.g. classifying images into 2 classes. Cross entropy measures the difference between two propbaility distributions and it is defined as:
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: Worth a mention that you should avoid adding a sigmoid at end of your network if using this loss with `from_sigmoid=True`, to avoid double sigmoid.

File: python/guide/packages/gluon/loss.md [hunk 2]
```diff
@@ -79,38 +83,79 @@ def plot(x, y):
...
-steps one-by-one, however, may need to numerical instabilities. The `loss`
+and then apply the cross entropy loss against the true labels: 
+
+$$ \begin{align}\begin{aligned}p = \softmax({pred})\\L = -\sum_i \sum_j {label}_j \log p_{ij}\end{aligned}\end{align}
+$$
+
+Running these two
+steps one-by-one, however, may lead to numerical instabilities. The `loss`
 module provides a single operators with softmax and cross entropy fused to avoid
 such problem.
```
Comment (thomelane) [inline python/guide/packages/gluon/loss.md]: Same warning as above: Worth a mention that you should avoid adding a sofmax at end of your network if using this loss with from_sigmoid=True, to avoid double softmax.

File: python/guide/packages/gluon/loss.md [hunk 3]
```diff
@@ -121,4 +166,120 @@ y = nd.array([0, 1])
 loss(x, y)
 ```
 
-For binary classification, we can use `SigmoidBinaryCrossEntropyLoss`.
+#### [Hinge Loss](https://mxnet.incubator.apache.org/api/python/gluon/loss.html#mxnet.gluon.loss.HingeLoss)
+
+Classifcation problems normally require a zero-one loss function, which assigns 0 loss to correct classifications and 1 otherwise. The problem of such a function is that it is hard to optimize and its gradients would be zero. Hinge Loss creates an upper bound on the zero-one loss function which makes it a convex and continous function.  It is defined as:
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: "Classifcation problems normally require a zero-one loss function, which assigns 0 loss to correct classifications and 1 otherwise." this isn't the case with cross entropy though?

[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: Can you give an example of when this is used.

File: python/guide/packages/gluon/loss.md [hunk 4]
```diff
@@ -121,4 +166,120 @@ y = nd.array([0, 1])
...
+$$
+The log loss decreases the closer the prediction is to the actual label. It is sensitive to outliers, because incorrectly classified points are penalized more.
+
+```python
+show_classification_loss(gloss.LogisticLoss())
+```
+
+#### [Kullback-Leibler Divergence Loss](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/loss.html#mxnet.gluon.loss.KLDivLoss)
+
+The Kullback-Leibler divergence loss measures the divergence between two propbaility distributions by calculating the difference between cross entropy and entropy. It takes as input the probaility of true label and the probabilty of predicted label.
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: true label and the probabilty of predicted label.  -> would switch the order of these since it's (pred, target)

File: python/guide/packages/gluon/loss.md [hunk 5]
```diff
@@ -121,4 +166,120 @@ y = nd.array([0, 1])
...
+print('output.softmax(): {}'.format(output.softmax().asnumpy().tolist()))
+target_dist = mx.nd.array([[0.3, 0.4, 0.1, 0.2]])
+loss_fn = mx.gluon.loss.KLDivLoss(from_logits=False)
+loss = loss_fn(output, target_dist)
+print('loss (kl divergence): {}'.format(loss.asnumpy().tolist()))
+```
+
+#### [Triplet Loss](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/loss.html#mxnet.gluon.loss.TripletLoss)
+
+Triplet loss takes three input tensors and measures the relative similarity. It takes a positive and negative input and the anchor.
```
Comment (thomelane) [inline python/guide/packages/gluon/loss.md]: tensors -> arrays

File: python/guide/packages/gluon/loss.md [hunk 6]
```diff
@@ -121,4 +166,120 @@ y = nd.array([0, 1])
...
+In the case of learning embeddings for images of characters, the network may get as input the following 3 images: 
+
+![triplet_loss](triplet_loss.png)
+
+The network would learn to minimize the distance between the two `A`'s and maximize the distance between `A` and `Z`.
+
+#### [CTC Loss](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/loss.html#mxnet.gluon.loss.CTCLoss)
+
+CTC Loss is the connectionist temporal classification loss. It is used to train recurrent neural networks with variable time dimension. It learns the alignment and labelling of input sequences. It takes a sequence as input and gives propabilies for each timestep. For instance in the followng image the word is not well algined with the 5 timesteps because of the different sizes of characters. CTC Loss finds for each timestep the highest propability e.g. `t1` presents with high propability a `C`. It combines the highest probapilities and returns the best path decoding. For an in-depth tutorial on how to use CTC-Loss in MXNet, check out this [example](https://github.com/apache/incubator-mxnet/tree/master/example/ctc).
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: Add link to https://distill.pub/2017/ctc/

File: python/guide/packages/gluon/loss.md [hunk 7]
```diff
@@ -121,4 +166,120 @@ y = nd.array([0, 1])
...
+#### [Cosine Embedding Loss](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/loss.html#mxnet.gluon.loss.CosineEmbeddingLoss)
+The cosine embedding loss computes the cosine distance between two input vectors. 
+
+$$
+\begin{split}L = \sum_i \begin{cases} 1 - {cos\_sim({input1}_i, {input2}_i)} & \text{ if } {label}_i = 1\\
+                 {cos\_sim({input1}_i, {input2}_i)} & \text{ if } {label}_i = -1 \end{cases}\\
+cos\_sim(input1, input2) = \frac{{input1}_i.{input2}_i}{||{input1}_i||.||{input2}_i||}\end{split}
+$$
+
+Cosine distance measures the similarity between two tensors given a label and is typically used for learning nonlinear embeddings.
```
Comment (thomelane) [inline python/guide/packages/gluon/loss.md]: tensors -> arrays

File: python/guide/packages/gluon/loss.md [hunk 8]
```diff
@@ -121,4 +166,120 @@ y = nd.array([0, 1])
...
+ - variance may not be constant
+
+Instead we can use a Poisson regression model, also known as log-linear model. Thereby the Poisson incident rate $$\mu$$ is 
+modelled by a linear combination of unknown parameters.
+We can then use the PoissonNLLLoss which calculates the negative log likelihood for a target that follows a Poisson distribution. 
+
+$$ L = \text{pred} - \text{target} * \log(\text{pred}) +\log(\text{target!}) $$
+
+
+
```
Comment (thomelane) [inline python/guide/packages/gluon/loss.md]: Add a conclusion and recommended next steps

File: python/guide/packages/gluon/loss.md [hunk 9]
```diff
@@ -13,6 +13,8 @@ from mxnet import nd, autograd
 from mxnet.gluon import nn, loss as gloss
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md [hunk 10]
```diff
@@ -62,7 +64,9 @@ l.backward()
 net.weight.grad()
 ```
 
-## Visualize Loss Functions
+## Loss functions
+
+Most commonly used loss functions can be divided into 2 classes: regression and classification. Regression loss functions output real-values, while classification loss functios output a class.
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md [hunk 11]
```diff
@@ -79,38 +83,79 @@ def plot(x, y):
...
+```python
+def show_classification_loss(loss):
+    x = nd.arange(-5, 5, .1)
+    y = loss(x, nd.ones_like(x))
+    plot(x, y) 
+```
+
+#### [L1 Loss](https://mxnet.incubator.apache.org/api/python/gluon/loss.html#mxnet.gluon.loss.L1Loss)
+
+L1Loss also called mean abolsute error computes the sum of absolute distance between target values and the output of the neural network. It is defined as:
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md [hunk 12]
```diff
@@ -79,38 +83,79 @@ def plot(x, y):
...
+
+It is a non-smooth function that can lead to non-convergence. It creates the same gradient for small and large loss values, which can be problematic for the learning process.
+
 ```{.python .input}
 show_regression_loss(gloss.L1Loss())
 ```
 
+#### [L2 Loss](https://mxnet.incubator.apache.org/api/python/gluon/loss.html#mxnet.gluon.loss.L2Loss)
+
+L2Loss also called Meaned Squared Error is a regression loss function that computes the squared distances between the target values and the output of the neural network. It is defined as:
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md [hunk 13]
```diff
@@ -79,38 +83,79 @@ def plot(x, y):
...
+Compared to L1, L2 loss it is a smooth function and it creates larger gradients for large loss values. However due to the squaring it puts high weight on outliers. 
+
 ```{.python .input}
 show_regression_loss(gloss.L2Loss())
 ```
 
-Then plot the classification losses with label values fixed to be 1.
+#### [Huber Loss](https://mxnet.incubator.apache.org/api/python/gluon/loss.html#mxnet.gluon.loss.HuberLosss)
+
+HuberLoss  combines advantages of L1 and L2 loss. It calculates a smoothed L1 loss that is equal to L1 if the adsolute error exceeds a threshold $$\rho$$, otherwise it is equal to L2. It is defined as:
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md [hunk 14]
```diff
@@ -121,4 +166,120 @@ y = nd.array([0, 1])
...
+
+#### [Kullback-Leibler Divergence Loss](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/loss.html#mxnet.gluon.loss.KLDivLoss)
+
+The Kullback-Leibler divergence loss measures the divergence between two propbaility distributions by calculating the difference between cross entropy and entropy. It takes as input the probaility of true label and the probabilty of predicted label. 
+
+$$
+L = \sum_i {label}_i * \big[\log({label}_i) - {pred}_i\big]
+$$
+
+The loss is large, if the prediced probality distribution is far from the groundtruth propability distribution. KL divergence is an asymmetric measure. KL divergence loss can be used in Variational Autoencoders (VAEs), and reinforcement learning policy networks such as Trust Region Policy Optimization (TRPO)
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md [hunk 15]
```diff
@@ -121,4 +166,120 @@ y = nd.array([0, 1])
...
+The cosine embedding loss computes the cosine distance between two input vectors. 
+
+$$
+\begin{split}L = \sum_i \begin{cases} 1 - {cos\_sim({input1}_i, {input2}_i)} & \text{ if } {label}_i = 1\\
+                 {cos\_sim({input1}_i, {input2}_i)} & \text{ if } {label}_i = -1 \end{cases}\\
+cos\_sim(input1, input2) = \frac{{input1}_i.{input2}_i}{||{input1}_i||.||{input2}_i||}\end{split}
+$$
+
+Cosine distance measures the similarity between two tensors given a label and is typically used for learning nonlinear embeddings.  
+For instance in the following code example we measure the similarity between the input vectors `x` and `y`. Since they are the same the label equals `1`. The loss function returns $$ \sum_i 1 - {cos\_sim({input1}_i, {input2}_i)} $$ which is equal `0`.
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md [hunk 16]
```diff
@@ -121,4 +166,120 @@ y = nd.array([0, 1])
...
+
+#### [PoissonNLLLoss](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/loss.html#mxnet.gluon.loss.PoissonNLLLoss)
+Poisson distribution is widely used for modelling count data. It is defined as:
+
+$$
+f(x) = \frac{\mu ^ {\kern 0.08 em x} e ^ {-\mu}} {x!} \qquad \qquad x = 0,1,2 , \ldots \,.
+$$
+
+
+For instance the count of cars in road traffic approximatly follows a Poisson distribution. Using an ordinary least squares model for Poisson distributed data would not work well because of two reasons:
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md [hunk 17]
```diff
@@ -121,4 +166,120 @@ y = nd.array([0, 1])
...
+The Kullback-Leibler divergence loss measures the divergence between two propbaility distributions by calculating the difference between cross entropy and entropy. It takes as input the probaility of true label and the probabilty of predicted label. 
+
+$$
+L = \sum_i {label}_i * \big[\log({label}_i) - {pred}_i\big]
+$$
+
+The loss is large, if the prediced probality distribution is far from the groundtruth propability distribution. KL divergence is an asymmetric measure. KL divergence loss can be used in Variational Autoencoders (VAEs), and reinforcement learning policy networks such as Trust Region Policy Optimization (TRPO)
+
+
+For instance in the following example we get a KL divergence of 0.02. We set ```from_logits=False```, so the loss functions will apply ```log_softmax``` on the network output, before computing the KL divergence.
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md (hunk 2 above)
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.

File: python/guide/packages/gluon/loss.md (hunk 7 above)
Comment (thomelane) [reply python/guide/packages/gluon/loss.md]: done.
